### DISCUSSION TITLE
Minimal example for custom app

### DISCUSSION THREAD

**Discussion by IngmarStein** [TechnitiumSoftware/DnsServer#1086] [Q&A]
https://blog.technitium.com/2021/03/creating-and-running-dns-apps-on.html has some pointers for building and packaging a custom app, but it requires Visual Studio and points at the built-in apps for examples. Is there a minimal, stand-alone example or template? I'm basically looking for environment (container?) where I could put some .cs files and run a command which packages the compiled result into a zip file ready for upload in the admin UI. Does that exist already or would you think of this as a worthy contribution?

**Comment by ShreyasZare** [comment]
Thanks for asking. 

> https://blog.technitium.com/2021/03/creating-and-running-dns-apps-on.html has some pointers for building and packaging a custom app, but it requires Visual Studio and points at the built-in apps for examples.

Yes, the old post mentions using Visual Studio for app development but its not really necessary. You can develop code any any editor and then use `dotnet publish` command to compile and publish binaries on all supported platforms. However, using Visual Studio for debugging complex code helps a lot.

> I'm basically looking for environment (container?) where I could put some .cs files and run a command which packages the compiled result into a zip file ready for upload in the admin UI. Does that exist already or would you think of this as a worthy contribution?

You can compile the project using `dotnet publish` command on a Linux setup and then zip to use it. The requirement is just to install .NET SDK on the platform you wish to compile. 

If you wish to use container for this you can use the [dotnet sdk](https://github.com/dotnet/dotnet-docker/blob/main/README.sdk.md) container published by Microsoft directly.

**Comment by IngmarStein** [comment]
I think I was looking for a bit more: a simple, standalone Hello World app which produces a zip file ready for upload.

"Compile using `dotnet publish` and then zip to use it" misses several key steps: 

- building the `TechnitiumLibrary` library and putting it in the right place
- getting the directory structure right to use DnsServerCore.ApplicationCommon
- adding `dnsApp.config` file to the zip file root

All of the above can be found be digging around and experimenting, but making this easier would lower the entry barrier for 3rd party devs to build an ecosystem around this project.

**Reply by ShreyasZare** [reply]
The project currently lacks basic documentation just because of lack of time, so a step by step tutorial on DNS app development wont be feasible for me at the moment. Which is why the blog post mentioned to take look at existing DNS apps to understand the concepts. Most of the apps are simple and have one or two class files in there so effectively they give a perspective on how to achieve similar task. The interfaces defined in ApplicationCommon project has all functions documented and explain how they work and what they return so that should give a fare idea on how to use them.

**Reply by IngmarStein** [reply]
I'm not looking for a step-by-step coding tutorial - I think the code-level documentation is sufficient.
More of a project setup which already comes with the required dependencies and build process. As you said, most apps are simple, so it would be nice to be able to put the one or two class files (and the config file) in a directory and have a build tool to produce a zip file.
It's not an issue for the built-in apps for which the infrastructure already exists and they are just built together with the main app.

**Reply by ShreyasZare** [reply]
I am not really sure what the exact expectations are here to be honest. Any of the existing DNS app project file gives the dependencies required. Someone developing a DNS app is expected to have Visual Studio setup where they just build the project and can zip the files manually from the output folder. If this requires automation for CI/CD scenario then the devops people already have enough experience to build and zip the project from source.

**Reply by IngmarStein** [reply]
Say you create a separate GitHub repository for your DNS app (and you want to build that without Visual Studio). What would be the recommended process to build that app? Clone it into an existing DnsServer workspace?

**Reply by ShreyasZare** [reply]
In that case, the DNS app can use the DLL binaries as dependencies directly and then it can be built independently. The DLL binaries can be used from the tar file provided for download or they can be built independently by cloning the library and the DNS server project.
